File: src/event_crop/src/event_crop_node.cpp
/*
 * C++ event cropping node — near-zero overhead.
 * Subscribes to EventArray, filters to ROI, publishes cropped events.
 */
#include <ros/ros.h>
#include <dv_ros_msgs/EventArray.h>

class EventCropNode {
public:
    EventCropNode(ros::NodeHandle& nh, ros::NodeHandle& pnh) {
        pnh.param("x_min", x_min_, 0);
        pnh.param("x_max", x_max_, 640);
        pnh.param("y_min", y_min_, 0);
        pnh.param("y_max", y_max_, 480);
        pnh.param("shift_coordinates", shift_, true);

        std::string input_topic, output_topic;
        pnh.param<std::string>("input_topic",  input_topic,  "/capture_node/events");
        pnh.param<std::string>("output_topic", output_topic, "/event_crop/events");

        int crop_w = x_max_ - x_min_;
        int crop_h = y_max_ - y_min_;
        ROS_INFO("Crop: x=[%d,%d) y=[%d,%d) -> %dx%d",
                 x_min_, x_max_, y_min_, y_max_, crop_w, crop_h);

        pub_ = nh.advertise<dv_ros_msgs::EventArray>(output_topic, 50);
        sub_ = nh.subscribe(input_topic, 50, &EventCropNode::callback, this);
    }

private:
    void callback(const dv_ros_msgs::EventArray::ConstPtr& msg) {
        if (msg->events.empty()) return;

        dv_ros_msgs::EventArray cropped;
        cropped.header = msg->header;
        cropped.width  = x_max_ - x_min_;
        cropped.height = y_max_ - y_min_;

        cropped.events.reserve(msg->events.size());

        for (const auto& e : msg->events) {
            if (e.x >= x_min_ && e.x < x_max_ &&
                e.y >= y_min_ && e.y < y_max_) {
                dv_ros_msgs::Event ce;
                if (shift_) {
                    ce.x = e.x - x_min_;
                    ce.y = e.y - y_min_;
                } else {
                    ce.x = e.x;
                    ce.y = e.y;
                }
                ce.ts = e.ts;
                ce.polarity = e.polarity;
                cropped.events.push_back(ce);
            }
        }

        if (!cropped.events.empty()) {
            pub_.publish(cropped);
        }
    }

    ros::Subscriber sub_;
    ros::Publisher  pub_;
    int x_min_, x_max_, y_min_, y_max_;
    bool shift_;
};

int main(int argc, char** argv) {
    ros::init(argc, argv, "event_crop_node");
    ros::NodeHandle nh;
    ros::NodeHandle pnh("~");
    EventCropNode node(nh, pnh);
    ros::spin();
    return 0;
}

File: src/event_crop/scripts/event_crop_node.py
#!/usr/bin/env python3
import rospy
from dv_ros_msgs.msg import EventArray, Event

class EventCropNode:
    def __init__(self):
        rospy.init_node('event_crop_node')

        self.x_min = rospy.get_param('~x_min', 0)
        self.x_max = rospy.get_param('~x_max', 100)
        self.y_min = rospy.get_param('~y_min', 0)
        self.y_max = rospy.get_param('~y_max', 100)

        input_topic  = rospy.get_param('~input_topic',  '/capture_node/events')
        output_topic = rospy.get_param('~output_topic', '/event_crop/events')
        self.shift_coordinates = rospy.get_param('~shift_coordinates', True)

        rospy.loginfo(f"Subscribing to {input_topic}")
        rospy.loginfo(f"Publishing cropped events to {output_topic}")
        rospy.loginfo(f"Cropping to x:[{self.x_min},{self.x_max}) y:[{self.y_min},{self.y_max})")

        self.pub = rospy.Publisher(output_topic, EventArray, queue_size=10)
        rospy.Subscriber(input_topic, EventArray, self.callback, queue_size=10)
        rospy.spin()

    def callback(self, msg):
        cropped = EventArray()
        cropped.header = msg.header
        cropped.width  = self.x_max - self.x_min
        cropped.height = self.y_max - self.y_min

        events = [
            e for e in msg.events
            if self.x_min <= e.x < self.x_max
            and self.y_min <= e.y < self.y_max
        ]

        if self.shift_coordinates:
            cropped.events = [
                Event(x=e.x - self.x_min, y=e.y - self.y_min, ts=e.ts, polarity=e.polarity)
                for e in events
            ]
        else:
            cropped.events = events

        if cropped.events:
            self.pub.publish(cropped)

if __name__ == '__main__':
    EventCropNode()
